x-pack/plugins/maps/server/content_management/maps_storage.ts

```diff
@@ -0,0 +1,316 @@
+    }
+
+    // Save data in DB
+    const soClient = await savedObjectClientFromRequest(ctx);
+    const savedObject = await soClient.create<MapSavedObjectAttributes>(
+      SO_TYPE,
+      dataToLatest,
+      optionsToLatest
+    );
+
+    // Validate DB response and DOWN transform to the request version
+    const { value, error: resultError } = transforms.create.out.result.down<
+      MapCreateOut,
+      MapCreateOut
+    >({
+      item: savedObjectToMapItem(savedObject, false),
+    });
+
+    if (resultError) {
+      throw Boom.badRequest(`Invalid response. ${resultError.message}`);
+    }
+
+    return value;
+  }
+
+  async update(
+    ctx: StorageContext,
+    id: string,
+    data: MapUpdateIn['data'],
+    options: UpdateOptions
+  ): Promise<MapUpdateOut> {
+    const {
+      utils: { getTransforms },
+      version: { request: requestVersion },
+    } = ctx;
+    const transforms = getTransforms(cmServicesDefinition, requestVersion);
+
+    // Validate input (data & options) & UP transform them to the latest version
+    const { value: dataToLatest, error: dataError } = transforms.update.in.data.up<
+      MapSavedObjectAttributes,
+      MapSavedObjectAttributes
+    >(data);
+    if (dataError) {
+      throw Boom.badRequest(`Invalid data. ${dataError.message}`);
+    }
+
+    const { value: optionsToLatest, error: optionsError } = transforms.update.in.options.up<
+      CreateOptions,
+      CreateOptions
+    >(options);
+    if (optionsError) {
+      throw Boom.badRequest(`Invalid options. ${optionsError.message}`);
+    }
+
+    // Save data in DB
+    const soClient = await savedObjectClientFromRequest(ctx);
+    const partialSavedObject = await soClient.update<MapSavedObjectAttributes>(
+      SO_TYPE,
+      id,
+      dataToLatest,
+      optionsToLatest
+    );
+
+    // Validate DB response and DOWN transform to the request version
+    const { value, error: resultError } = transforms.update.out.result.down<
+      MapUpdateOut,
+      MapUpdateOut
+    >({
+      item: savedObjectToMapItem(partialSavedObject, true),
+    });
+
+    if (resultError) {
+      throw Boom.badRequest(`Invalid response. ${resultError.message}`);
+    }
+
+    return value;
+  }
+
+  async delete(ctx: StorageContext, id: string): Promise<MapDeleteOut> {
+    const soClient = await savedObjectClientFromRequest(ctx);
+    await soClient.delete(SO_TYPE, id);
+    return { success: true };
+  }
+
+  async search(
+    ctx: StorageContext,
+    query: SearchQuery,
+    options: MapSearchOptions = {}
+  ): Promise<MapSearchOut> {
+    const {
+      utils: { getTransforms },
+      version: { request: requestVersion },
+    } = ctx;
+    const transforms = getTransforms(cmServicesDefinition, requestVersion);
+    const soClient = await savedObjectClientFromRequest(ctx);
+
+    // Validate and UP transform the options
+    const { value: optionsToLatest, error: optionsError } = transforms.search.in.options.up<
```

Comment:
Should there be a similar process for the `query` object?

Reply:
The `query` has a concrete interface. You can't add anything to it. Maybe your need to pass an additional setting to your search layer, that's what `Options` object are for.